### PR TITLE
Memoize Environment.Value() to improve performance

### PR DESCRIPTION
**Pull request by grossag**
The code I work on calls env.Value() often on the same values as part of consolidating outside dependencies. This change improves performance of repeated call by memoizing the results.

While at it, I fixed several flake8 validation issues in Python.py and PythonTests.py.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by mwichmann** [comment]
There's an existing memoization framework, would it makes sense to hook into that?

**Comment by grossag** [comment]
> There's an existing memoization framework, would it makes sense to hook into that?

I am open to doing that but am having trouble figuring out how that would work in practice. I am assuming that you are talking about more than just adding the @SCons.Memoize.CountMethodCall decorator tag, right?

I considered doing memoization only inside of env.Value(), so basically not changing Python.py at all. However, that would not allow us to support memoization in ValueNodeInfo.str_to_node().

File: src/engine/SCons/Node/Python.py [hunk 1]
```diff
@@ -173,6 +176,23 @@ def get_csig(self, calc=None):
         self.get_ninfo().csig = contents
         return contents
 
+
+def ValueWithMemo(value, built_value=None):
+    global _memoLookupMap
+
+    # No current support for memoizing a value that needs to be built.
+    if built_value:
```
Comment (bdbaddog) [inline src/engine/SCons/Node/Python.py:184]: Can you add a test to cover this as well?

Comment (grossag) [reply src/engine/SCons/Node/Python.py:184]: Done, thanks!

File: src/engine/SCons/Node/PythonTests.py [hunk 2]
```diff
@@ -98,18 +97,43 @@ def test_get_csig(self):
...
+        assert value1 is value2
+
+        # Next confirm that ValueNodeInfo.str_to_node does memoization using
+        # the same cache as ValueWithMemo.
+        ni = SCons.Node.Python.ValueNodeInfo()
+        value3 = ni.str_to_node('vvv')
+        assert value1 is value3
+
+
+class BuiltValueNoMemoTestCase(unittest.TestCase):
```
Comment (bdbaddog) [inline src/engine/SCons/Node/PythonTests.py]: Pretty sure you can actually move this to test_built_value(self) in the ValueMemoTestCase

**Comment by bdbaddog** [comment]
Why keep the original Value() ?

**Comment by grossag** [comment]
@bdbaddog So consumers aren’t broken. I wanted to add memoization in a transparent way to consumers. The two things I didn’t want to break are: (1) usage of env.Value() and (2) usage of isinstance(node, SCons.Node.Python.Value).

**Comment by dragon512** [comment]
So is this not about adding a namespace for the Value node? ( like we have for File/Dir with the fs object and Alias nodes with the Alias namespace

I like the idea, but I think it changes some behavior in that calling

```python
v = env.Value("a")
v1 = env.Value("a") 
```
v == v1 -> False

currently makes two different nodes. This changes the code to allow v == v1 to be the same node. I think that is a good thing BTW.

I should point out that there is a possible logic bug I think you missed in the tests ( I might be wrong).. ie

```python

env["foo"] = "hello"
env2["foo"] = "world"

v=env.Value("$foo")
v1=env2.Value("$foo")

# as these are different environments that have different final values
v != v1

```

**Comment by grossag** [comment]
@dragon512 Can you explain more about why v should be not equal to v1? What about them actually make them different values? I had thought that Value instances which are not built (built_value=None) are simply the equivalent of strings. I don't see any calls to subst() in that class and get_text_contents() simply returns the string value.

**Comment by bdbaddog** [comment]
@dragon512  - since `env.Value("$SOME_VAR")` is a literal and not evaluated, then `env2.Value('$SOME_VAR')` can be the same value right?

**Comment by dragon512** [comment]
Sorry about the delay here I needed to make a local test to double-check my understanding.

@grossag is correct in that Value() only holds the value given.. no subst() call on the value. 

So what I stated in not 100% correct given the current state.

This patch would change the behavior of Value() to return a common object based on the value in the Value object. This means that while the current code would to this:

v=env.Value("$foo")
v1=env1.Value("$foo")
v2=env2.Value("$foo")

(v != v1 !=v2) -> True

The patch would change this to:

(v == v1 == v2) -> True

I am for this change. This would save memory and speed up the build for cases in which values are used as sources to a builder. (Values as targets are currently has some issues.. but that is a different problem)

I like the addition to have the built_value return a different Value node with the value of built_value

A note we probably need to add to the documentation to explain the case of:

env.Value(["hello","world"]) != env.Value(["hello","world"]) -> True

I have seen cases in which Value was used to pass List or Dictionary objects. The example above could be a hickup that might surprise some people that can be fixed by

mylist = ["hello","world"]
env.Value(mylist) == env.Value(mylist) -> True

File: src/engine/SCons/Node/Python.py [hunk 3]
```diff
@@ -173,6 +176,23 @@ def get_csig(self, calc=None):
...
+        return Value(value, built_value)
+
+    value_str = str(value)
```
Comment (bdbaddog) [inline src/engine/SCons/Node/Python.py]: Should this be wrapped with try/except in the case the value can't be converted to str ?

Comment (grossag) [reply src/engine/SCons/Node/Python.py]: I don't see the __str__ function using try/except so AFAICT this class assumes that the value can be converted to a string. I don't mind wrapping in try/except but I would want to add a test case. Do you know of any examples that I should try to support?

Comment (bdbaddog) [reply src/engine/SCons/Node/Python.py]: If value is mutable then it shouldn't be cached right? (list, dict, other?)

Comment (grossag) [reply src/engine/SCons/Node/Python.py]: So you are saying that you want me to limit this to primitive types: int, bool, float, str?

Comment (grossag) [reply src/engine/SCons/Node/Python.py]: Ok I limited to primitive types and pushed an update.

**Comment by bdbaddog** [comment]
I wasn't specifically instructing a change.
I was asking if it should be limited to immutable types?
Would that impact the performance improvement for your usage?

**Comment by grossag** [comment]
My usage is entirely limited to passing strings. They aren’t fixed hardcoded strings, but instead are dynamically generated then passed into the function. I don’t know whether they would be considered immutable but I also don’t know of a way in python to check for immutable objects.

**Comment by mwichmann** [comment]
Strings are always immutable...

**Comment by grossag** [comment]
Ok thanks. In light of that, changing to only memoizing immutable objects will still provide the performance benefit I’m looking for.

**Comment by dragon512** [comment]
There are cases in which people pass dictionaries. The textfile and substfile builders are examples of this. I think this should be supported as well as this could break stuff. However, I strongly agree we want items to be immutable. I feel in this case of collections these get copied or something to make them unique and maybe stringify the collection to allow it to work in with the hash lookup.

**Comment by bdbaddog** [comment]
@dragon512 - is caching mutable objects ever correct? You're caching them because you want the same return value when presented with the same input. But if it's mutable, the cost of comparing previous calls input values with current may remove any benefit from caching. So simpler to only cache when immutable.

**Comment by bdbaddog** [comment]
> Ok thanks. In light of that, changing to only memoizing immutable objects will still provide the performance benefit I’m looking for.

Still not what I'm asking.  I'm trying to make sure that whatever is done is correct. And secondly in your usage if you're already caching mutable objects how much hit would not caching them in this logic present?

So unless someone convinces me otherwise. I'd say we should not be caching mutable objects, only immutable objects.

Looking at this:
https://stackoverflow.com/questions/8056130/immutable-vs-mutable-types

It seems a reasonable argument that types which don't have __hash__ are safe to consider immutable. (Obeying Python's general usage of duck typing).

This should be handled by logic to catch this I'd think:
```
>>> x={}
>>> y={}
>>> y[x] = 'blah'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
>>> y['blah']='blah'
```

I'm guessing `try/except TypeError` should be sufficient.

**Comment by mwichmann** [comment]
> @dragon512 - is caching mutable objects ever correct? 

sure it can be correct, but if what you want is to *hash* the object, then it has to be immutable, cause else they wouldn't has the same. That's why it has to be immutable to be a dict key.

**Comment by dragon512** [comment]
So is this a cache or a namespace? ( I know we are saying a cache, but I feel these have to handle as File objects or Aliases ) The issue with values is if they reuse or create a new node.

We will get odd behaviors if people can say:

env.Value("foo") == env.Value("foo") -> true
x = {1:2}
env.Value(x) == env.Value(x) -> false

I feel that allowing this is only going to cause confusion and broken builds. 

I agree we want items to be immutable. Large dictionaries/sets/lists can be an issue to make immutable. it becomes an issue of fast vs correct at some point. These can become immutable in that we make a (deep) copy in the collection and that can be hashed based on value ( which we tend to have most of the code for this in Scons already to generate md5 sigs). The collections also have to contain other collections or objects that are immutable.

I would argue that we could handle the collection and basic immutables in values. custom objects have to have a hash function. A document that collections are handled in a safe way, but can cause a speed concern ( more so if you have large collections). 

I would then have not allow values to accept non-hashable objects. ( I would have this be an error)

I just feel that we want to avoid breaking anything. I know values are used with dictionaries already. However, these cases only work today correctly when we deal with the dictionaries correctly handled as immutables.

**Comment by bdbaddog** [comment]
@dragon512 
Nope. not a namespace. 
Yes. Only cache immutable values. The rest would get the current Value() object.
If you check Value==Value for two different invocations today it fails.. 
Are you checking that today?

If the object being cached provides a __hash__ method then it'll work.
I don't see any issues assuming it behaves as I've described above.

**Comment by dragon512** [comment]
If you don't see an issue. I guess it is good. This should be a good improvement to the system.

I see an issue only in that the other nodes don't behave like this and it is a behavior that I find people expect (File/DIr/Entity and Alias have a namespace). That fact that Values don't seemed like a bug, not a feature. I do see value however in adding a new node type that can hold values and behave in a different way. However, that is beyond the scope of this PR.

**Pull request merged**